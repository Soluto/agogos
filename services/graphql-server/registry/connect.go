package registry

import (
	agogos "agogos/generated"
	"context"
	"io"
	"os"
	"time"

	log "github.com/sirupsen/logrus"
	"google.golang.org/grpc"
)

var (
	address = os.Getenv("REGISTRY_URL")
)

func closeWithContext(ctx context.Context, cls io.Closer) {
	<-ctx.Done()
	err := cls.Close()
	if err != nil {
		log.WithError(err).Warningln("Error closing")
	}
}

func createSubscribeClient(ctx context.Context) (agogos.Registry_SubscribeClient, error) {
	log.Infoln("Connecting to registry")
	conn, err := grpc.DialContext(ctx, address, grpc.WithInsecure())
	if err != nil {
		log.WithError(err).Warn("Error initiating GRPC channel")
		return nil, err
	}
	log.Infoln("Registry connection initiated")

	registryClient := agogos.NewRegistryClient(conn)
	stream, err := registryClient.Subscribe(ctx, &agogos.SubscribeParams{})
	if err != nil {
		log.WithError(err).Warn("Error subscribing to registry", err)
		conn.Close()
		return nil, err
	}

	log.Infoln("Subscribed to registry")

	go closeWithContext(ctx, conn)

	return stream, nil
}

func Connect(ctx context.Context) (chan *agogos.ConfigurationMessage, chan error) {
	configCh := make(chan *agogos.ConfigurationMessage)
	errCh := make(chan error)

	go func() {
		for {
			var err error
			clientCtx, cancel := context.WithCancel(ctx)
			client, err := createSubscribeClient(clientCtx)
			if err == nil {
				for {
					var config *agogos.ConfigurationMessage
					config, err = client.Recv()

					if err == nil {
						configCh <- config
					} else {
						break
					}
				}
			}

			errCh <- err
			cancel()
			time.Sleep(10 * time.Second)
		}
	}()

	return configCh, errCh
}
