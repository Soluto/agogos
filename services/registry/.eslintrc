{
    "env": {
        "es6": true,
        "mocha": true,
        "jest/globals": true
    },
    "extends": [
        "standard"
    ],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module"
    },
    "plugins": [
        "@typescript-eslint",
        "jest"
    ],
    "rules": {
        "@typescript-eslint/no-unused-vars": "error",
        "space-before-function-paren": "off",
        "space-in-parens": "off",
        "array-bracket-spacing": "off",
        "semi": "off",
        "@typescript-eslint/semi": ["error"],
        "jest/no-disabled-tests": "warn",
        "jest/no-focused-tests": "error",
        "jest/no-identical-title": "error",
        "jest/prefer-to-have-length": "warn",
        "jest/valid-expect": "error"
    },
    "overrides": [{
        "files": ["tests/**"],
        "rules": {
            "no-unused-expressions": "off"
        }
    }]
}