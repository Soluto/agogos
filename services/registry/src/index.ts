import nconf = require("nconf");
import logger, { loggingMiddleware } from "./logger";
nconf.env("__");

import * as express from "express";
import gqlSchemaRoute from "./routes/schema";
import validateRoute from "./routes/validate";
import { startGrpcServer } from "./sync/sync-grpc-server";

const PORT = process.env.PORT || 4000;

const app = express();

app.use(loggingMiddleware);

app.get("/health", (_: express.Request, res: express.Response) =>
    res.send(true),
);
app.get("/metrics", (_: express.Request, res: express.Response) =>
    res.send(true),
);

app.use("/schema", gqlSchemaRoute);
app.use("/validate", validateRoute);

app.listen({ port: PORT }, () =>
    logger.info(`ðŸš€ HTTP Server ready at http://localhost:${PORT}`),
);

startGrpcServer();
