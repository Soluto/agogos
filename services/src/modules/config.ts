import * as path from 'path';
import * as envVar from 'env-var';
import { AuthenticationConfig } from './authentication/types';

const envVarExt = envVar.from(process.env, {
  asSet: (value: string) => new Set(value.split(',')),
});

// General
export const httpPort = envVarExt.get('PORT').default('8080').asIntPositive();
export const logLevel = envVarExt.get('LOG_LEVEL').default('WARN').asString();
export const nodeEnv = envVarExt.get('NODE_ENV').default('development').asString();

// Repositories
export const useS3ResourceRepository = envVarExt.get('USE_S3_RESOURCE_REPOSITORY').default('false').asBoolStrict();
export const useFileSystemResourceRepository = envVarExt
  .get('USE_FS_RESOURCE_REPOSITORY')
  .default('true')
  .asBoolStrict();

// Resources
export const resourceUpdateInterval = envVarExt.get('RESOURCE_UPDATE_INTERVAL').default('60000').asIntPositive();

// GraphQL configuration
export const enableGraphQLTracing = envVarExt.get('GRAPHQL_TRACING').default('true').asBool();
export const enableGraphQLPlayground = envVarExt.get('GRAPHQL_PLAYGROUND').default('true').asBoolStrict();
export const enableGraphQLIntrospection = envVarExt.get('GRAPHQL_INTROSPECTION').default('true').asBoolStrict();

// Policies
export const tmpPoliciesDir = envVarExt
  .get('TMP_POLICIES_DIR')
  .default(path.resolve(process.cwd(), 'tmp/policies'))
  .asString();

export const ignorePolicies =
  envVar.get('NODE_ENV').required().asString() !== 'production' &&
  envVar.get('IGNORE_POLICIES').default('false').asBoolStrict();

// Authentication
const defaultAuthenticationConfig: AuthenticationConfig = {
  anonymous: {
    publicPaths: ['/metrics', '/.well-known/apollo/server-health', '/graphql'],
  },
};
export const authenticationConfig =
  (envVarExt.get('AUTHENTICATION_CONFIGURATION').asJsonObject() as AuthenticationConfig) ?? defaultAuthenticationConfig;

// Plugins
export const pluginsDir = envVar.get('PLUGINS_DIR').asString();
export const pluginsConfig = envVar.get('PLUGINS_CONFIGURATION').asJsonObject() as Record<string, unknown> | undefined;

// TODO:Check the list of apollo federation directives when upgrading apollo version
const defaultKnownApolloDirectives = 'key,extends,external,requires,provides';
export const knownApolloDirectives = envVarExt
  .get('KNOWN_APOLLO_DIRECTIVES')
  .default(defaultKnownApolloDirectives)
  .asSet();
