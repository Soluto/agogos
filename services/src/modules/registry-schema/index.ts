import typeDefs from './typedefs';
import resolvers from './resolvers';
export * from './types';

export { typeDefs, resolvers };
