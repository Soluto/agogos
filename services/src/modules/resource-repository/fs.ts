import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
