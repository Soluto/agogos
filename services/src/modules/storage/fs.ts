import * as path from 'path';
import { promises as fs } from 'fs';
import pLimit from 'p-limit';
import { Storage, FileStats, listFilesItem } from '.';

export class FileSystemStorage implements Storage {
  async fileStats(filePath: string): Promise<FileStats> {
    const stats = await fs.stat(filePath);

    return { lastModified: stats.mtime };
  }

  async readFile(filePath: string): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { [k in any]: never }): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  async readFile(filePath: string, options: { asString?: boolean } = {}): Promise<{ content: Buffer | string }> {
    if (options.asString) return { content: await fs.readFile(filePath, 'utf8') };

    return { content: await fs.readFile(filePath) };
  }

  async writeFile(filePath: string, content: string | Buffer): Promise<void> {
    await fs.writeFile(filePath, content);
  }

  async deleteFile(filePath: string): Promise<void> {
    await fs.unlink(filePath);
  }

  async listFiles(folderPath: string): Promise<listFilesItem[]> {
    const filenames = await fs.readdir(folderPath);
    const limit = pLimit(10);

    return Promise.all(
      filenames.map(filename => {
        return limit(async () => {
          const filePath = path.resolve(folderPath, filename);
          const { lastModified } = await this.fileStats(filePath);

          return { filePath, lastModified };
        });
      })
    );
  }

  async mkdir(folderPath: string, options?: { recursive: true }): Promise<void> {
    await fs.mkdir(folderPath, options);
  }
}
