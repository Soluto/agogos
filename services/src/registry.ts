import {ApolloServer, gql, IResolvers} from 'apollo-server-fastify';
import * as fastify from 'fastify';
import pLimit from 'p-limit';

import {fetchAll, updateAll, ResourceGroup, applyResourceGroupUpdates} from './modules/resource-repository';
import * as config from './modules/config';
import {validateResourceGroupOrThrow} from './modules/validation';
import logger from './modules/logger';
import {handleSignals, handleUncaughtErrors} from './modules/shutdownHandler';

const typeDefs = gql`
    # General
    input ResourceMetadataInput {
        namespace: String!
        name: String!
    }

    type Result {
        success: Boolean!
    }

    input ResourceGroupInput {
        schemas: [SchemaInput!]
        upstreams: [UpstreamInput!]
        upstreamClientCredentials: [UpstreamClientCredentialsInput!]
    }

    type Query {
        validateResourceGroup(input: ResourceGroupInput!): Result
        validateSchemas(input: [SchemaInput!]!): Result
        validateUpstreams(input: [UpstreamInput!]!): Result
        validateUpstreamClientCredentials(input: [UpstreamClientCredentialsInput!]!): Result
    }

    type Mutation {
        updateResourceGroup(input: ResourceGroupInput!): Result
        updateSchemas(input: [SchemaInput!]!): Result
        updateUpstreams(input: [UpstreamInput!]!): Result
        updateUpstreamClientCredentials(input: [UpstreamClientCredentialsInput!]!): Result
    }

    # Schemas
    input SchemaInput {
        metadata: ResourceMetadataInput!
        schema: String!
    }

    # Upstreams
    enum AuthType {
        ActiveDirectory
    }

    """
    GraphQL doesn't support unions for input types, otherwise this would be a union of different auth types.
    Instead, the AuthType enum indicates which auth type is needed, and there's a property which corresponds to each auth type, which we validate in the registry.
    """
    input AuthInput {
        type: AuthType!
        activeDirectory: ActiveDirectoryAuthInput!
    }

    input ActiveDirectoryAuthInput {
        authority: String!
        resource: String!
    }

    input UpstreamInput {
        metadata: ResourceMetadataInput!
        host: String!
        auth: AuthInput!
    }

    # Upstream client credentials
    input ActiveDirectoryCredentials {
        authority: String!
        clientId: String!
        clientSecret: String!
    }

    """
    GraphQL doesn't support unions for input types, otherwise this would be a union of different auth types.
    Instead, the AuthType enum indicates which auth type is needed, and there's a property which corresponds to each auth type, which we validate in the registry.
    """
    input UpstreamClientCredentialsInput {
        metadata: ResourceMetadataInput!
        authType: AuthType!
        activeDirectory: ActiveDirectoryCredentials!
    }
`;

interface ResourceMetadataInput {
    namespace: string;
    name: string;
}

interface ResourceGroupInput {
    schemas?: SchemaInput[];
    upstreams?: UpstreamInput[];
    upstreamClientCredentials?: UpstreamClientCredentialsInput[];
}

interface SchemaInput {
    metadata: ResourceMetadataInput;
    schema: string;
}

export enum AuthType {
    ActiveDirectory = 'ActiveDirectory',
}

interface ActiveDirectoryAuthInput {
    authority: string;
    resource: string;
}

interface UpstreamInput {
    metadata: ResourceMetadataInput;
    host: string;
    auth: {
        type: AuthType;
        activeDirectory: ActiveDirectoryAuthInput;
    };
}

interface UpstreamClientCredentialsInput {
    metadata: ResourceMetadataInput;
    authType: AuthType;
    activeDirectory: {
        authority: string;
        clientId: string;
        clientSecret: string;
    };
}

async function fetchAndValidate(updates: Partial<ResourceGroup>): Promise<ResourceGroup> {
    const rg = await fetchAll();
    const newRg = applyResourceGroupUpdates(rg, updates);
    validateResourceGroupOrThrow(newRg);

    return newRg;
}

const singleton = pLimit(1);
const resolvers: IResolvers = {
    Query: {
        async validateResourceGroup(_, args: {input: ResourceGroupInput}) {
            await fetchAndValidate(args.input);

            return {success: true};
        },
        async validateSchemas(_, args: {input: SchemaInput[]}) {
            await fetchAndValidate({schemas: args.input});

            return {success: true};
        },
        async validateUpstreams(_, args: {input: UpstreamInput[]}) {
            await fetchAndValidate({upstreams: args.input});

            return {success: true};
        },
        async validateUpstreamClientCredentials(_, args: {input: UpstreamClientCredentialsInput[]}) {
            await fetchAndValidate({upstreamClientCredentials: args.input});

            return {success: true};
        },
    },
    Mutation: {
        updateResourceGroup(_, args: {input: ResourceGroupInput}) {
            return singleton(async () => {
                const rg = await fetchAndValidate(args.input);
                await updateAll(rg);

                return {success: true};
            });
        },
        updateSchemas(_, args: {input: SchemaInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({schemas: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
        updateUpstreams(_, args: {input: UpstreamInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({upstreams: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
        updateUpstreamClientCredentials(_, args: {input: UpstreamClientCredentialsInput[]}) {
            return singleton(async () => {
                const rg = await fetchAndValidate({upstreamClientCredentials: args.input});
                await updateAll(rg);

                return {success: true};
            });
        },
    },
};

export const app = new ApolloServer({
    typeDefs,
    resolvers,
    tracing: config.enableGraphQLTracing,
    playground: config.enableGraphQLPlayground,
    introspection: config.enableGraphQLIntrospection,
});

async function run() {
    logger.info('Stitch registry booting up...');
    const server = fastify();
    server.register(app.createHandler({path: '/graphql'}));
    const address = await server.listen(config.httpPort, '0.0.0.0');
    logger.info({address}, 'Stitch registry started');

    handleSignals(() => app.stop());
    handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
    run();
}
