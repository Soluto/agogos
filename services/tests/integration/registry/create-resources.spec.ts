import { exec } from 'child_process';
import * as path from 'path';
import { promises as fs } from 'fs';
import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import { gql } from 'apollo-server-core';
import { mocked } from 'ts-jest/utils';
import * as nock from 'nock';
import { beforeEachDispose } from '../before-each-dispose';
import { app } from '../../../src/registry';
import { mockResourceBucket } from '../resource-bucket';
import {
  DefaultUpstream,
  PolicyType,
  PolicyDefinition,
  ResourceGroup,
  Schema,
  Upstream,
  UpstreamClientCredentials,
} from '../../../src/modules/resource-repository';
import { tmpPoliciesDir } from '../../../src/modules/config';
import mockFsForOpa from '../../helpers/mock-fs-for-opa';
import { AuthType } from '../../../src/modules/registry-schema';
import { getOpaBuildWasmCommand } from '../../../src/modules/opa-helper';

jest.mock('child_process', () => ({
  exec: jest.fn((_, cb) => cb()),
}));

const mockedExec = mocked(exec, true);

const schema: Schema = {
  metadata: { namespace: 'namespace', name: 'name' },
  schema: 'type Query { something: String! }',
};

const upstream: Upstream = {
  metadata: { namespace: 'namespace', name: 'name' },
  host: 'test.api',
  auth: {
    type: AuthType.ActiveDirectory,
    activeDirectory: { authority: 'https://authority', resource: 'someResource' },
  },
};

const defaultUpstream: DefaultUpstream = {
  targetOrigin: 'http://localhost:8080',
  headers: [
    {
      name: 'x-api-client',
      value: '{incomingRequest?.headers?.["x-api-client"] ?? "Unknown"}',
    },
  ],
};

const upstreamClientCredentials: UpstreamClientCredentials = {
  metadata: { namespace: 'namespace', name: 'name' },
  authType: AuthType.ActiveDirectory,
  activeDirectory: {
    authority: 'https://authority',
    clientId: 'myClientId',
    clientSecret: 'myClientSecret',
  },
};

const policy: PolicyDefinition = {
  metadata: { namespace: 'namespace', name: 'name' },
  type: PolicyType.opa,
  code: `real rego code
           with multiple
           lines`,
  args: {
    an: { type: 'String', default: '{source.an}' },
    another: { type: 'String!' },
  },
  query: {
    gql: 'some gql',
    variables: {
      a: 'b',
    },
  },
};

const baseResourceGroup: ResourceGroup = {
  schemas: [],
  upstreams: [],
  upstreamClientCredentials: [],
  policies: [],
};

describe('Create resource', () => {
  let client: ApolloServerTestClient;
  let bucketContents: { current: ResourceGroup; policyFiles: { [name: string]: string } };

  beforeEachDispose(() => {
    client = createTestClient(app);
    bucketContents = mockResourceBucket(baseResourceGroup);

    return () => nock.cleanAll();
  });

  it('creates a Schema', async () => {
    const response = await client.mutate({
      mutation: gql`
        mutation CreateSchema($schema: SchemaInput!) {
          updateSchemas(input: [$schema]) {
            success
          }
        }
      `,
      variables: {
        schema,
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updateSchemas: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();
  });

  it('creates a Schema using updateResourceGroup', async () => {
    const response = await client.mutate({
      mutation: gql`
        mutation UploadResources($resourceGroup: ResourceGroupInput!) {
          updateResourceGroup(input: $resourceGroup) {
            success
          }
        }
      `,
      variables: {
        resourceGroup: { schemas: [schema] },
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updateResourceGroup: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();
  });

  it('Upstream', async () => {
    const response = await client.mutate({
      mutation: gql`
        mutation CreateUpstream($upstream: UpstreamInput!) {
          updateUpstreams(input: [$upstream]) {
            success
          }
        }
      `,
      variables: {
        upstream,
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updateUpstreams: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();
  });

  it('Default upstream', async () => {
    const response = await client.mutate({
      mutation: gql`
        mutation SetDefaultUpstream($defaultUpstream: DefaultUpstreamInput!) {
          setDefaultUpstream(input: $defaultUpstream) {
            success
          }
        }
      `,
      variables: {
        defaultUpstream,
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ setDefaultUpstream: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();
  });

  it('UpstreamClientCredentials', async () => {
    const response = await client.mutate({
      mutation: gql`
        mutation CreateUpstreamClientCredentials($upstreamClientCredentials: UpstreamClientCredentialsInput!) {
          updateUpstreamClientCredentials(input: [$upstreamClientCredentials]) {
            success
          }
        }
      `,
      variables: {
        upstreamClientCredentials,
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updateUpstreamClientCredentials: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();
  });

  it('creates an opa type policy', async () => {
    mockFsForOpa.mock();

    const response = await client.mutate({
      mutation: gql`
        mutation CreatePolicy($policy: PolicyInput!) {
          updatePolicies(input: [$policy]) {
            success
          }
        }
      `,
      variables: {
        policy,
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updatePolicies: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();

    const compiledFilename = 'namespace-name.wasm';
    const uncompiledPath = path.resolve(tmpPoliciesDir, 'namespace-name.rego');
    const compiledPath = path.resolve(tmpPoliciesDir, compiledFilename);
    const regoCode = `package policy\n${policy.code}`;
    expect(fs.writeFile).toHaveBeenCalledWith(uncompiledPath, regoCode);
    expect(fs.unlink).toHaveBeenCalledWith(uncompiledPath);
    expect(fs.unlink).toHaveBeenCalledWith(compiledPath);
    expect(fs.readFile).toHaveBeenCalledWith(compiledPath);
    expect(bucketContents.policyFiles).toEqual({ [compiledFilename]: 'compiled rego code' });

    const expectedCommand = getOpaBuildWasmCommand(uncompiledPath, compiledPath);
    expect(mockedExec.mock.calls[0][0]).toBe(expectedCommand);

    mockFsForOpa.restore();
  });

  it('creates an opa type policy using updateResourceGroup', async () => {
    mockFsForOpa.mock();

    const response = await client.mutate({
      mutation: gql`
        mutation CreatePolicy($resourceGroup: ResourceGroupInput!) {
          updateResourceGroup(input: $resourceGroup) {
            success
          }
        }
      `,
      variables: {
        resourceGroup: { policies: [policy] },
      },
    });

    expect(response.errors).toBeUndefined();
    expect(response.data).toEqual({ updateResourceGroup: { success: true } });
    expect(bucketContents.current).toMatchSnapshot();

    const compiledFilename = 'namespace-name.wasm';
    const uncompiledPath = path.resolve(tmpPoliciesDir, 'namespace-name.rego');
    const compiledPath = path.resolve(tmpPoliciesDir, compiledFilename);
    const regoCode = `package policy\n${policy.code}`;
    expect(fs.writeFile).toHaveBeenCalledWith(uncompiledPath, regoCode);
    expect(fs.unlink).toHaveBeenCalledWith(uncompiledPath);
    expect(fs.unlink).toHaveBeenCalledWith(compiledPath);
    expect(fs.readFile).toHaveBeenCalledWith(compiledPath);
    expect(bucketContents.policyFiles).toEqual({ [compiledFilename]: 'compiled rego code' });

    const expectedCommand = getOpaBuildWasmCommand(uncompiledPath, compiledPath);
    expect(mockedExec.mock.calls[0][0]).toBe(expectedCommand);

    mockFsForOpa.restore();
  });
});
