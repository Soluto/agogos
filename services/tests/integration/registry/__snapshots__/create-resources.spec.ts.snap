// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create resource Default upstream 1`] = `
Object {
  "defaultUpstream": Object {
    "headers": Array [
      Object {
        "name": "x-api-client",
        "value": "{incomingRequest?.headers?.[\\"x-api-client\\"] ?? \\"Unknown\\"}",
      },
    ],
    "targetOrigin": "http://localhost:8080",
  },
  "pluginsData": Object {},
  "policies": Array [],
  "policyAttachments": Object {},
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Create resource Upstream 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [],
  "policyAttachments": Object {},
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [
    Object {
      "auth": Object {
        "activeDirectory": Object {
          "authority": "https://authority",
          "resource": "someResource",
        },
        "type": "ActiveDirectory",
      },
      "host": "test.api",
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
    },
  ],
}
`;

exports[`Create resource UpstreamClientCredentials 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [],
  "policyAttachments": Object {},
  "schemas": Array [],
  "upstreamClientCredentials": Array [
    Object {
      "activeDirectory": Object {
        "authority": "https://authority",
        "clientId": "myClientId",
        "clientSecret": "myClientSecret",
      },
      "authType": "ActiveDirectory",
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
    },
  ],
  "upstreams": Array [],
}
`;

exports[`Create resource creates a Schema 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [],
  "policyAttachments": Object {},
  "schemas": Array [
    Object {
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
      "schema": "type Query { something: String! }",
    },
  ],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Create resource creates a Schema using updateResourceGroup 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [],
  "policyAttachments": Object {},
  "schemas": Array [
    Object {
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
      "schema": "type Query { something: String! }",
    },
  ],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Create resource creates an opa type policy 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [
    Object {
      "args": Object {
        "an": Object {
          "default": "{source.an}",
          "optional": false,
          "type": "String",
        },
        "another": Object {
          "optional": false,
          "type": "String!",
        },
      },
      "code": "real rego code
           with multiple
           lines",
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
      "query": Object {
        "gql": "some gql",
        "variables": Object {
          "a": "b",
        },
      },
      "type": "opa",
    },
  ],
  "policyAttachments": Object {},
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;

exports[`Create resource creates an opa type policy using updateResourceGroup 1`] = `
Object {
  "pluginsData": Object {},
  "policies": Array [
    Object {
      "args": Object {
        "an": Object {
          "default": "{source.an}",
          "optional": false,
          "type": "String",
        },
        "another": Object {
          "optional": false,
          "type": "String!",
        },
      },
      "code": "real rego code
           with multiple
           lines",
      "metadata": Object {
        "name": "name",
        "namespace": "namespace",
      },
      "query": Object {
        "gql": "some gql",
        "variables": Object {
          "a": "b",
        },
      },
      "type": "opa",
    },
  ],
  "policyAttachments": Object {},
  "schemas": Array [],
  "upstreamClientCredentials": Array [],
  "upstreams": Array [],
}
`;
